System.Diagnostics.TraceSource unit tests

MarkPflug opened the pull request:
Unit tests for System.Diagnostics.TraceSource. Only marginally novel thing to note is the idea of abstract unit tests for Collection types; see TraceListenerCollectionClassTests.cs. I know these non-generic collections were the norm pre .net 2.0. So, being able to have a common set of tests for all IList implementations might be a quick way to get coverage on these types.

I noticed that the API still exposes Refresh calls that were supposed to update trace settings from configuration. However, there is no configuration anymore. Should the APIs be removed? Is configuration something that is coming back in the future? I vaguely remember hearing that the XML configuration was being replaced by some format-agnostic mechanism.

MarkPflug commented:
This should cover issue https://github.com/dotnet/corefx/issues/925.

Maxwe11 commented on src/System.Diagnostics.TraceSource/tests/TraceClassTests.cs:
```diff
@@ -1,122 +1,373 @@
+            Trace.Listeners.Add(listener);
+            Trace.WriteLine((Object)"Text");
+            listener.Flush();
+            Assert.Equal("Text\r\n", listener.Output);
```
Will this fail under Linux/OSX ?

MarkPflug replied:
I suspect you are right. I tried to write this as Environment.NewLine but the compiler was complaining about resolving Environment for some reason. I meant to come back and investigate further, but I guess I forgot. The two existing unit tests had hard-coded 
"/r/n" as well, not that that is an excuse.

MarkPflug commented:
... and apparently I need to figure out why the tests are failing on the CI build.

Maxwe11 commented:
@MarkPflug it failing because
    
	TraceListenerCollectionClassTests.cs(71,21): warning CS0114: 'System.Diagnostics.TraceSourceTests.TraceListenerCollectionClassTests.AddTest()' hides inherited member 'System.Diagnostics.TraceSourceTests.ListBaseTests<System.Diagnostics.TraceListenerCollection>.AddTest()'. To make the current member override that implementation, add the override keyword. Otherwise add the new keyword. [d:\j\workspace\dotnet_corefx_linux_debug_prtest\src\System.Diagnostics.TraceSource\tests\System.Diagnostics.TraceSource.Tests.csproj]
	TraceListenerCollectionClassTests.cs(124,21): warning CS0114: 'System.Diagnostics.TraceSourceTests.TraceListenerCollectionClassTests.CopyToTest()' hides inherited member 'System.Diagnostics.TraceSourceTests.CollectionBaseTests<System.Diagnostics.TraceListenerCollection>.CopyToTest()'. To make the current member override that implementation, add the override keyword. Otherwise add the new keyword. [d:\j\workspace\dotnet_corefx_linux_debug_prtest\src\System.Diagnostics.TraceSource\tests\System.Diagnostics.TraceSource.Tests.csproj]
	TraceListenerCollectionClassTests.cs(136,21): warning CS0114: 'System.Diagnostics.TraceSourceTests.TraceListenerCollectionClassTests.IndexOfTest()' hides inherited member 'System.Diagnostics.TraceSourceTests.ListBaseTests<System.Diagnostics.TraceListenerCollection>.IndexOfTest()'. To make the current member override that implementation, add the override keyword. Otherwise add the new keyword. [d:\j\workspace\dotnet_corefx_linux_debug_prtest\src\System.Diagnostics.TraceSource\tests\System.Diagnostics.TraceSource.Tests.csproj]
	TraceListenerCollectionClassTests.cs(153,21): warning CS0114: 'System.Diagnostics.TraceSourceTests.TraceListenerCollectionClassTests.InsertTest()' hides inherited member 'System.Diagnostics.TraceSourceTests.ListBaseTests<System.Diagnostics.TraceListenerCollection>.InsertTest()'. To make the current member override that implementation, add the override keyword. Otherwise add the new keyword. [d:\j\workspace\dotnet_corefx_linux_debug_prtest\src\System.Diagnostics.TraceSource\tests\System.Diagnostics.TraceSource.Tests.csproj]
	TraceListenerCollectionClassTests.cs(164,21): warning CS0114: 'System.Diagnostics.TraceSourceTests.TraceListenerCollectionClassTests.RemoveTest()' hides inherited member 'System.Diagnostics.TraceSourceTests.ListBaseTests<System.Diagnostics.TraceListenerCollection>.RemoveTest()'. To make the current member override that implementation, add the override keyword. Otherwise add the new keyword. [d:\j\workspace\dotnet_corefx_linux_debug_prtest\src\System.Diagnostics.TraceSource\tests\System.Diagnostics.TraceSource.Tests.csproj]
	TraceEventCacheClassTests.cs(28,36): error CS0117: 'Interop' does not contain a definition for 'mincore' [d:\j\workspace\dotnet_corefx_linux_debug_prtest\src\System.Diagnostics.TraceSource\tests\System.Diagnostics.TraceSource.Tests.csproj]

MarkPflug commented:
I've encountered an interesting issue on my local machine. When I run the tests I get a lot of failures due to:
System.InvalidOperationException : Collection was modified; enumeration operation may not execute.
at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
 at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
 at System.Collections.Generic.List`1.Enumerator.MoveNext()
...

However, these only fail when I run without coverage.
MSBuild /T:Test 
lots of failures

MSBuild /T:Test /p:Coverage=true
no failures.

Has anyone seen this before? It is very reproducible, but I am unable to debug it because I only have VS 2013.

stephentoub commented:
By default, xunit runs test classes in parallel, but we disable that for code coverage runs. You're likely hitting race conditions been your test classes as a result, either due to your own shared state or that in the library, e.g. one class with a test case that enumerates Trace.Listeners and another class with a case that modifies Trace.Listeners. Given the design of the library, you probably want to disable parallelism for all runs, e.g. https://github.com/dotnet/corefx/blob/master/src/System.Linq.Parallel/tests/XunitAssemblyAttributes.cs. Separately we should consider whether there are things we could do to improve the thread-safety of the library.

MarkPflug commented:
Ah, that would explain it. I suspect this library will require disabled parallelism even if we do improve the thread-safety. The Trace class is static and stores all of its state in static fields, it isn't going to play well with parallel tests. I suspect that the particular issue that I ran into is that when the UseGlobalLock flag is disabled, then there is literally no lock taken when enumerating the attached listeners. Clearly the flag is intended to avoid contention on that lock, but it seems like there should at least be a reader/writer lock used to prevent Listeners collection being modified by a different thread. Probably not an issue that would be encountered in the wild since Tracing is typically only initialized through configuration and not actually modified at runtime.

stephentoub commented on src/System.Diagnostics.TraceSource/src/Properties/InternalsVisibleTo.cs:
```diff
@@ -0,0 +1,9 @@
+
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("System.Diagnostics.TraceSource.Tests, PublicKey=002400000480000094000000060200000024000052534131000400000100010007d1fa57c4aed9f0a32e84aa0faefd0de9e8fd6aec8f87fb03766c834c99921eb23be79ad9d5dcc1dd9ad236132102900b723cf980957fc4e177108fc607774f29e8320e92ea05ece4e821c0a5efe8f1645c4c0c93c1ab99285d622caa652c1dfad63d745d6f2de5f17e5eaf0fc4963d261c8a12436518206dc093344d5ad293 ...
```
How important is this for the tests?  While we haven't always succeeded, in general we strive not to give the tests internals access to the source, even if it means sacrificing some coverage.

weshaggard replied:
+1 for trying to avoid this.

MarkPflug replied:
The two classes that are impacted are TraceInternal and TraceListenerCollection. TraceInternal can be mostly tested through Trace. TraceListenerCollection has an internal constructor, so options would be to either allow internals, not test it, or use reflection to create instances of it. Would using reflection to construct it be preferred to internals visible to?

stephentoub commented on src/System.Diagnostics.TraceSource/tests/TraceClassTests.cs:
```diff
@@ -1,122 +1,375 @@
+            Trace.UseGlobalLock = DiagnosticsConfiguration.UseGlobalLock;
+            Trace.AutoFlush = DiagnosticsConfiguration.AutoFlush;
+            Trace.Listeners.Clear();
+            Trace.Listeners.Add(new DefaultTraceListener());
```
Should all of this resetting also be done in a Dispose on the tests class?

stephentoub commented on src/System.Diagnostics.TraceSource/tests/TraceClassTests.cs:
```diff
@@ -1,122 +1,375 @@
+            Trace.UseGlobalLock = true;
+            Assert.True(Trace.UseGlobalLock);
+            Trace.UseGlobalLock = false;
+            Assert.False(Trace.UseGlobalLock);
```
Should each individual test try to restore the world to the state before the test, i.e. Trace.UseGlobalLock should be reset at the end to whatever it was at the beginning?

MarkPflug replied:
The intent was that after every test the default state would be restored to the Trace class. As you mentioned above, this could be done by implementing IDisposable on all the test classes and calling TraceTestHelper.ResetState().

stephentoub commented on src/System.Diagnostics.TraceSource/tests/TraceEventCacheClassTests.cs:
```diff
@@ -0,0 +1,52 @@
+            var id = cache.ProcessId;
+            // TODO: This was causing a compile error on the linux CI build. 
+            // (I don't have an environment to replicate this in)
+            //var expected = Interop.mincore.GetCurrentProcessId();
```
You do, actually :smile: Just pass /p:OS=Linux to msbuild when building, e.g. in your tests directory do "msbuild /t:rebuild /t:BuildAndTest /p:OS=Linux".

The problem is that GetCurrentProcessId is a P/Invoke to a Win32 function, and it's defined in a file that's only included when building for Windows, i.e. /p:OS=Windows_NT.  If you want to validate that cache.ProcessId is the ID of the current process, you can update the packages.config for the test assembly to include System.Diagnostics.Process, and then here you can use Process.GetCurrentProcess().Id.  That'll be available and work on Linux.

stephentoub commented on src/System.Diagnostics.TraceSource/tests/TraceTestHelper.cs:
```diff
@@ -0,0 +1,40 @@
+                return new StringBuilder().AppendLine().ToString();
+                // NOTE: Want to use this line:                
+                //return System.Environment.NewLine;
+                // However, I get this compile error, that I have absoltely no idea how to resolve.
```
You can add System.Runtime.Extensions to the packages.config, essentially adding a reference to System.Runtime.Extensions.dll, in order to get access to System.Environment.

stephentoub commented:
Thanks so much for writing all of these tests, @MarkPflug.  I left a few comments... once address, this LGTM.

weshaggard commented:
cc @karelz anyone you would like to review changes like these in the tracing APIs?

MarkPflug commented:
@stephentoub, I think I addressed all the issues you pointed out. I need to delete some code that I left commented out, and it appears have tabs enabled for XML that I need to fix. I refactored the TraceInternal tests to invoke through the Trace class instead (to drop internals visible to), but I should rename the test class back to 'TraceInternal', since that is the class that it really being covered by those tests. I should be able to get touch this up later this evening.

stephentoub commented:
@dotnet-bot test this please

stephentoub commented:
Thanks, @MarkPflug.  This all looks good.  Once you've fixed those issues you mentioned (e.g. tabs), we can go ahead and merge this in.  Again, thanks for the contributions!

stephentoub commented:
Thanks, @MarkPflug! LGTM.
